observers/qos_correction.hpp

```diff
@@ -0,0 +1,143 @@
+#ifndef SERENITY_QOS_CORRECTION_HPP
+#define SERENITY_QOS_CORRECTION_HPP
+
+#include <list>
+#include <vector>
+
+#include "glog/logging.h"
+
+#include "mesos/mesos.hpp"
+#include "mesos/resources.hpp"
+
+#include "mesos/slave/oversubscription.hpp"
+
+#include "messages/serenity.hpp"
+
+#include "serenity/serenity.hpp"
+
+namespace mesos {
+namespace serenity {
+
+inline std::list<ResourceUsage_Executor> filterPrExecutors(
+    ResourceUsage usage) {
+  std::list<ResourceUsage_Executor> beExecutors;
+  for (ResourceUsage_Executor inExec : usage.executors()) {
+    if (!inExec.has_executor_info()) {
+      LOG(ERROR) << "Executor <unknown>"
+      << " does not include executor_info";
+      // Filter out these executors.
+      continue;
+    }
+    if (inExec.allocated().size() == 0) {
+      LOG(ERROR) << "Executor "
+      << inExec.executor_info().executor_id().value()
+      << " does not include allocated resources.";
+      // Filter out these executors.
+      continue;
+    }
+
+    Resources allocated(inExec.allocated());
+    // Check if task uses revocable resources.
+    if (!allocated.revocable().empty())
+      beExecutors.push_back(inExec);
+  }
+
+  return beExecutors;
+}
+
+
+using ContentionDeciderFunction = Try<QoSCorrections>
+    (const Contentions& currentContentions,
+     const ResourceUsage& currentUsage);
+
+
+/**
+ * Convenient base class for contention interpretations.
+ * It converts contentions & usage to QoSCorrections.
+ *
+ * Convenient for debugging and testing different algorithms.
+ */
+class ContentionDecider {
+ public:
+  virtual ContentionDeciderFunction decide = 0;
+};
+
+
+/**
+ * Checks contentions and choose executors to kill.
+ * Currently it takes sorted contentions by severity.
+ * We are iterating from the most serious contention to the less serious
+ * and choose BE executors to kill.
+ * WARNING(bplotka): This interpreter algorithm assumes that severity is
+ * reflecting how many resources is needed for PR job to not starve again.
+ */
+class SeverityBasedCpuDecider : public ContentionDecider {
+ public:
+  ContentionDeciderFunction decide;
+};
+
+
+/**
+ * QoSCorrectionObserver observes incoming Contentions and
+ * ResourceUsage and produces QoSCorrections.
+ *
+ * Currently Slave understand only kill correction action.
+ */
+class QoSCorrectionObserver : public MultiConsumer<Contentions>,
+                              public Consumer<ResourceUsage>,
```

Comment:
In future you should synchronize ResourceUsage inside SyncConsumer (currently MultiConsumer) class.